contracts/teller-network/MetadataStore.sol

```diff
@@ -80,28 +79,69 @@ contract MetadataStore is Ownable {
         arbitration = _arbitration;
     }
 
-    function setEscrowAddress(address _escrow) public onlyOwner {
-        escrow = _escrow;
+    mapping(address => uint) public user_nonce;
+
+    function getDataHash(
+        string calldata _username,
+        bytes calldata _statusContactCode
+    ) external view returns (bytes32) {
+        return dataHash(_username, _statusContactCode, user_nonce[msg.sender]);
     }
 
-    function setLicense(address _license) public onlyOwner {
-        license = _license;
+    function dataHash(
+        string memory _username,
+        bytes memory _statusContactCode,
+        uint nonce
+    ) internal view returns (bytes32) {
+        return keccak256(abi.encodePacked(address(this), _username, _statusContactCode, nonce));
     }
 
-    function setArbitration(address _arbitration) public onlyOwner {
-        arbitration = _arbitration;
+    function getSigner(
+        string memory _username,
+        bytes memory _statusContactCode,
+        uint nonce,
+        bytes memory _signature
+    ) internal view returns(address) {
+        return recoverAddress(getSignHash(dataHash(_username, _statusContactCode, nonce)), _signature);
+    }
+
+    function getMessageSigner(
+        string calldata _username,
+        bytes calldata _statusContactCode,
+        bytes calldata _signature,
+        uint _nonce
+    ) external view returns(address) {
+        return getSigner(_username, _statusContactCode, _nonce, _signature);
+    }
+
+    function addOrUpdateUser(
+        bytes memory _signature,
+        bytes memory _statusContactCode,
+        string memory _location,
+        string memory _username,
+        uint _nonce
+    ) public returns(address payable _user) {
+        _user = address(uint160(getSigner(_username, _statusContactCode, _nonce, _signature)));
+        require(_nonce == user_nonce[_user], "Invalid nonce");
+        user_nonce[_user]++;
+        _addOrUpdateUser(_user, _statusContactCode, _location, _username);
+        return _user;
     }
 
     function addOrUpdateUser(
-        address _user,
         bytes memory _statusContactCode,
         string memory _location,
         string memory _username
     ) public {
-        if(msg.sender != escrow){
-            require(msg.sender == _user, "Sender does not match address");
-        }
+        _addOrUpdateUser(msg.sender, _statusContactCode, _location, _username);
+    }
```

Comment:
the differences are:
- `function addOrUpdateUser( bytes memory _signature, bytes memory _statusContactCode,         string memory _location, string memory _username, uint _nonce)` will obtain the user address from the signature and call `_addOrUpdateUser`
- `function addOrUpdateUser( bytes memory _statusContactCode,         string memory _location, string memory _username, uint _nonce)`  since it doesn't receive an user, it will use `msg.sender` when calling `_addOrUpdateUser`.  It's mostly for convenience, since calling this function will not require obtaining a signature, if the user to be added/updated is the msg.sender.